aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,307 @@
+    super.initialize(properties);
+    this.dynamo = AwsClientFactories.from(properties).dynamo();
+    this.lockTableName = properties.get(CatalogProperties.LOCK_TABLE);
+    Preconditions.checkNotNull(lockTableName, "DynamoDB lock table name must not be null");
+    ensureLockTableExistsOrCreate();
+  }
+
+  @Override
+  public boolean acquire(String entityId, String ownerId) {
+    try {
+      Tasks.foreach(entityId)
+          .throwFailureWhenFinished()
+          .retry(Integer.MAX_VALUE - 1)
+          .exponentialBackoff(acquireIntervalMs(), acquireIntervalMs(), acquireTimeoutMs(), 1)
+          .onlyRetryOn(
+              ConditionalCheckFailedException.class,
+              ProvisionedThroughputExceededException.class,
+              TransactionConflictException.class,
+              RequestLimitExceededException.class,
+              InternalServerErrorException.class)
+          .run(id -> acquireOnce(id, ownerId));
+      return true;
+    } catch (DynamoDbException e) {
+      return false;
+    }
+  }
+
+  @VisibleForTesting
+  void acquireOnce(String entityId, String ownerId) {
+    GetItemResponse response = dynamo.getItem(GetItemRequest.builder()
+        .tableName(lockTableName)
+        .key(toKey(entityId))
+        .build());
+
+    if (!response.hasItem()) {
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST)
+          .build());
+    } else {
+      Map<String, AttributeValue> currentItem = response.item();
+
+      try {
+        Thread.sleep(Long.parseLong(currentItem.get(COL_LEASE_DURATION_MS).n()));
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(
+            String.format("Fail to acquire lock %s by %s, interrupted during sleep", entityId, ownerId), e);
+      }
+
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST_OR_VERSION_MATCH)
+          .expressionAttributeValues(ImmutableMap.of(
+              ":eid", AttributeValue.builder().s(entityId).build(),
+              ":vid", AttributeValue.builder().s(currentItem.get(COL_VERSION).s()).build()))
+          .build());
+    }
+
+    startNewHeartbeat(entityId, ownerId);
+  }
+
+  private void startNewHeartbeat(String entityId, String ownerId) {
+    if (heartbeats.containsKey(entityId)) {
+      heartbeats.remove(entityId).cancel(false);
+    }
+
+    heartbeats.put(entityId, scheduler().scheduleAtFixedRate(() -> dynamo.putItem(PutItemRequest.builder()
+        .tableName(lockTableName)
+        .item(toNewItem(entityId, ownerId))
+        .conditionExpression(CONDITION_LOCK_ID_MATCH)
+        .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+        .build()), 0, heartbeatIntervalMs(), TimeUnit.MILLISECONDS));
```

Comment:
I think this task needs to handle exceptions and keep running.

If there is a connection issue that causes this to fail, then the task will not be rescheduled. Instead, the future will contain the error. Then other concurrent processes could hold the lock even though this process thinks it holds the lock.

At a minimum, I think that this thread should log any exception thrown by dynamo and keep trying to update its lease.

It would be great if we could also detect if the lock is no longer held (any put fails with `ConditionalCheckFailedException`) and stop any commit attempts, but I don't see a way to do that. This should definitely log an error if `ConditionalCheckFailedException` is thrown and note that two processes may think they both hold the lock.

Reply:
Yes, that is definitely a risk. I also thought about what you described, but I think the only way to achieve that is to allow the lock manager caller to explicitly call a `heartbeat` method to emit heartbeat, and this method needs to be passed all the way to every place of a commit to ensure heartbeat does not timeout.

So comparing with the current Hive lock, we are basically choosing between:
1. if we lock without heartbeat, a super long timeout is needed to ensure the commit must succeed, and if lock dies, all processes would be blocked for that long time.
2. if we lock with heartbeat, we don't need a long timeout. But we need to either choose to:
    1. heartbeat in the background, which makes the interface more decoupled, but would lead to this issue of potential idle heartbeat thread
    2. heartbeat by the caller, which would require changes the commit interface all over the place to pass the heartbeat around.

I think so far 2.1 is still the best approach to go for now, since we have already made sure `release` must be called after a commit, so heartbeat will be killed. If it is not called, the process must have crashed and the heartbeat thread would also be dead.